File: AtCoder/KyoproTenke/python/002-Encyclopedia_of_Parentheses.py
from sys import setrecursionlimit
setrecursionlimit(10**9)


N = int(input())


def dfs(left, right, s):
    if left == 0 and right == 0:
        print(s)
        return
    if left > 0:
        dfs(left - 1, right, s + '(')
    if right > left:
        dfs(left, right - 1, s + ')')


if N % 2 == 0:
    dfs(N // 2, N // 2, '')

File: AtCoder/KyoproTenkei/cpp/002-Encyclopedia_of_Parentheses.cpp
#include <bits/stdc++.h>
using namespace std;
using ll = long long;

ll N;

void dfs(ll left, ll right, string s) {
  if (left == 0 && right == 0) {
    puts(s.c_str());
    return;
  }
  if (left > 0) dfs(left - 1, right, s + '(');
  if (right > left) dfs(left, right - 1, s + ')');
}

int main() {
  cin >> N;
  if (N % 2 == 0) dfs(N / 2, N / 2, "");
}